README
= Project: bigbluebutton-api-ruby

== Goal

Provide access to the BigBlueButton web conferencing API.

== Dependencies

Tested in Ruby 1.8.7 and Ruby 1.9.2.

Needs: rake, rubygems, nokogiri

== Releases

=== Version 0.0.4
Gem renamed from 'bigbluebutton' to 'bigbluebutton-api-ruby'.

First version with support also to BBB API version 0.7.

Changes:

- API functions now return a hash and instead of the XML returned by BBB. The XML is converted to a hash that uses symbols as keys and groups keys with the same name.

=== Version 0.0.3
Fixes module issue preventing proper throwing of exceptions.

=== Version 0.0.1
This is the first version of this gem. It provides an implementation of the 0.64 bbb API, with the following exceptions:

- Does not implement meeting token, and instead relies on meeting id as the unique identifier for a meeting.
- Documentation suggests there is way to call join_meeting as API call (instead of browser URL). This call currently does not work as documented.

== Contact

=== Version 0.0.4+

Author::     Leonardo Crauss Daronco
Email::      leonardodaronco@gmail.com
Project::    GT-Mconf: Multiconference system for interoperable web and mobile @ PRAV Labs - UFRGS
Home Page::  http://www.inf.ufrgs.br/prav/gtmconf/
License::    Distributed under same terms as Ruby

=== Version 0.0.3 and below

Author::     Joe Kinsella
Email::      joe.kinsella@gmail.com
Home Page::  http://www.brownbaglunch.com/bigbluebutton
License::    Distributes under same terms as Ruby

CHANGELOG.md
# Change Log

## [1.9.1] - 2023-03-31
* [#60] Prevent problems with double slashes on API calls [thanks to [@farhatahmad](https://github.com/farhatahmad)]

[#60]: https://github.com/mconf/bigbluebutton-api-ruby/pull/60
[1.9.1]: https://github.com/mconf/bigbluebutton-api-ruby/compare/v1.9.0...v1.9.1

## [1.9.0] - 2022-05-03
* [#56] Add support for checksum using SHA256

[#56]: https://github.com/mconf/bigbluebutton-api-ruby/pull/56
[1.9.0]: https://github.com/mconf/bigbluebutton-api-ruby/compare/v1.8.0...v1.9.0

## [1.8.0] - 2021-12-06
* [#43] Add keys to every `BigBlueButtonException`, to better identify them.
* [#42] Change `BigBlueButtonException` to inherit from `StandardError` instead of `Exception`.
* [#50] [BREAKING-CHANGE] Replace `debug` flag with a optional Logger. The application using the
  gem can pass its own logger as argument for the `BigBlueButtonApi` initialization.
  If none is passed, the gem will use its own default logger on `STDOUT` with `INFO` level.
* [#40] Fix issue preventing documents from being preuploaded using the create call.
* Fix parse of recordings with invalid times. It would break at `getRecordings` if one
  of them had an empty `startTime` or `endTime`.
* Add a Dockerfile and compose to help run tests.
* Fix deprecated `TimeoutError` constant.
* [#35] Make `get_recordings` accept multiple `state` params.
* [#34] Update `rubyzip` gem to the newest version with no vulnerability, from 1.2.2 to 1.3.0
* [#33] Upgrade dependencies:
  - `childprocess` from 0.3.2 to 1.0.1
  - `ffi` from 1.0.11 to 1.9.24
  - `json` from 1.8.3 to 1.8.6
  - `nokogiri` from 1.6.6.2 to 1.10.4
  - `rack` from 1.4.1 to 1.6.11
  - `rdoc` from 3.12 to 3.12.1
  - `rubyzip` from 0.9.8 to 1.2.2

## [1.7.0] - 2018-08-17

* [#29] Add support to the API call `updateRecordings` introduced in BigBlueButton 1.1.
* [#31] Fixed issue with length=nil breaking multiple recording formats.
* Call `setConfigXML` via POST and change encoding method. Fixes issues with special
  characters (such as `*`) in the config.xml.
* Add method to return the URL to `/check`.

## [1.6.0] - 2016-06-15

* Rename BigBlueButtonApi#salt to #secret

## [1.5.0] - 2016-04-07

* Add 1.0 as a supported version of BigBlueButton.
* [#1686] Automatically set the version number of a server by fetching it from
its API.
* [#1686] Fix comparison of config.xml strings that would sometimes thing XMLs
were different in cases when they were not.
* [#1695] Add support for servers that use HTTPS.

## [1.4.0] - 2015-07-20

* Updated default ruby version to 2.2.0.
* Add support for BigBlueButton 0.9 (includes all 0.9.x). Consists only in
  accepting the version "0.9", since this new API version doesn't break
  compatibility with the previous version and also doesn't add new API
  calls.

## [1.3.0] - 2014-05-11

* Drop support to BigBlueButton 0.7 and add BigBlueButton to 0.81. Includes
  support to the new API calls `getDefaultConfigXML` and `setConfigXML`.
* Reviewed all methods responses (some changed a bit) and documentation.
* Allow non standard options to be passed to **all** API calls.
* Removed method `join_meeting` that was usually not used. Should always use
  `join_meeting_url`.
* Moved the hash extension method `from_xml` to its own class to prevent
  conflicts with Rails. See
  https://github.com/mconf/bigbluebutton-api-ruby/pull/6.

## [1.2.0] - 2013-03-13

*  Allow non standard options to be passed to some API calls. These API calls are: create_meeting, join_meeting_url, join_meeting, get_recordings.
  Useful for development of for custom versions of BigBlueButton.
* Accept :record as boolean in create_meeting
* Better formatting of data returned by get_recordings

## [1.1.1] - 2013-01-30

* BigBlueButtonApi can now receive http headers to be sent in all get/post
  requests

## [1.1.0] - 2012-05-04

* Updated ruby to 1.9.3-194.
* Support to BigBlueButton 0.4 rc1.

## [1.0.0] - 2012-05-04

* Version 0.1.0 renamed to 1.0.0.

## [0.1.0] - 2011-11-25

* Support to BigBlueButton 0.8:
  * New methods for recordings: get_recordings, publish_recordings,
    delete_recordings
  * Pre-upload of slides in create_meeting
  * New parameters added in the already existent methods
  * For more information see BigBlueButton docs at
    http://code.google.com/p/bigbluebutton/wiki/API#Version_0.8
* Method signature changes: create_meeting, join_meeting_url and
  join_meeting. Optional parameters are now passed using a hash.
* Integration tests for the entire library using cucumber.
* Changed the XML parser to xml-simple (especially to solve issues with
  CDATA values).

## [0.0.11] - 2011-09-01

* The file "bigbluebutton-api" was renamed to "bigbluebutton_api". All
  "require" calls must be updated.
* Splitted the library in more files (more modular) and created rspec tests
  for it.
* Added a BigBlueButtonApi.timeout attribute to timeout get requests and
  avoid blocks when the server is down. Defaults to 2 secs.
* New method last_http_response to access the last HTTP response object.
* Automatically detects the BBB server version if not informed by the user.

## [0.0.10] - 2011-04-28

* Returning hash now will **always** have these 3 values: :returncode
  (boolean), :messageKey (string) and :message (string).
* Some values in the hash are now converted to a fixed variable type to
  avoid inconsistencies:
  * :meetingID (string)
  * :attendeePW (string)
  * :moderatorPW (string)
  * :running (boolean)
  * :hasBeenForciblyEnded (boolean)
  * :endTime and :startTime (DateTime or nil)

## [0.0.9] - 2011-04-08

* Simplied "attendees" part of the hash returned in get_meeting_info. Same
  thing done for get_meetings.

## [0.0.8] - 2011-04-06

* New method get_api_version that returns the version of the server API (>= 0.7).
* New simplified hash syntax for get_meetings. See docs for details.

## [0.0.7] - 2011-04-06

## [0.0.6] - 2011-04-05

* New method test_connection.
* Added comparison method for APIs.

lib/bigbluebutton-api.rb
require 'net/http'
require 'cgi'
require 'rexml/document'
require 'digest/sha1'
require 'rubygems'
require 'nokogiri'
require 'hash_to_xml'

module BigBlueButton

  class BigBlueButtonException < Exception
    attr_accessor :key

    def to_s
      unless key.blank?
        super.to_s + ", messageKey: #{key}"
      else
        super
      end
    end

  end

  # This class provides access to the BigBlueButton API. BigBlueButton
  # is an open source project that provides web conferencing for distance
  # education (http://code.google.com/p/bigbluebutton/wiki/API). This API
  # was developed to support the following version of BBB: 0.7, 0.8 (soon)
  #
  # Sample usage of the API is as follows:
  # 1) Create a meeting with the create_meeting call
  # 2) Direct a user to either join_meeting_url
  # 3) To force meeting to end, call end_meeting
  #
  # 0.0.4+:
  # Author::    Leonardo Crauss Daronco  (mailto:leonardodaronco@gmail.com)
  # Copyright:: Copyright (c) 2011 Leonardo Crauss Daronco
  # Project::   GT-Mconf: Multiconference system for interoperable web and mobile @ PRAV Labs - UFRGS
  # License::   Distributes under same terms as Ruby
  #
  # 0.0.3 and below:
  # Author::    Joe Kinsella  (mailto:joe.kinsella@gmail.com)
  # Copyright:: Copyright (c) 2010 Joe Kinsella
  # License::   Distributes under same terms as Ruby
  #
  # Considerations about the returning hash:
  # * The XML returned by BBB is converted to a Hash. See the desired method's documentation for examples.
  # * Three values will *always* exist in the hash: :returncode (boolean), :messageKey (string) and :message (string)
  # * Some of the values returned by BBB are converted to better represent the data. Some of these are listed
  #   bellow. They will *always* have the type informed:
  #   * :meetingID (string)
  #   * :attendeePW (string)
  #   * :moderatorPW (string)
  #   * :running (boolean)
  #   * :hasBeenForciblyEnded (boolean)
  #   * :endTime and :startTime (DateTime or nil)
  #
  class BigBlueButtonApi

    attr_accessor :url, :supported_versions, :salt, :version, :debug, :timeout

    # Initializes an instance
    # url::       URL to a BigBlueButton server (e.g. http://demo.bigbluebutton.org/bigbluebutton/api)
    # salt::      Secret salt for this server
    # version::   API version: 0.7 (valid for 0.7, 0.71 and 0.71a)
    def initialize(url, salt, version='0.7', debug=false)
      @supported_versions = ['0.7']
      @url = url
      @salt = salt
      @debug = debug
      @timeout = 2 # 2 seconds timeout for get requests

      if version.nil?
        @version = get_api_version
      else
        @version = version
      end
      unless @supported_versions.include?(@version)
        raise BigBlueButtonException.new("BigBlueButton error: Invalid API version #{version}. Supported versions: #{@supported_versions.join(', ')}")
      end

      puts "BigBlueButtonAPI: Using version #{@version}" if @debug
    end

    # Returns the url used to join the meeting
    # meeting_id::        Unique identifier for the meeting
    # user_name::         Name of the user
    # password::          Password for this meeting - used to set the user as moderator or attendee
    # user_id::           Unique identifier for this user
    # web_voice_conf::    Custom voice-extension for users using VoIP
    def join_meeting_url(meeting_id, user_name, password,
                         user_id = nil, web_voice_conf = nil)

      params = { :meetingID => meeting_id, :password => password, :fullName => user_name,
                 :userID => user_id, :webVoiceConf => web_voice_conf }
      get_url(:join, params)
    end

    # Creates a new meeting. Returns the hash with the response or
    # throws BigBlueButtonException on failure.
    # meeting_name::        Name for the meeting
    # meeting_id::          Unique identifier for the meeting
    # moderator_password::  Moderator password
    # attendee_password::   Attendee password
    # welcome_message::     Welcome message to display in chat window
    # dialin_number::       Dial in number for conference using a regular phone
    # logout_url::          URL to return user to after exiting meeting
    # voice_bridge::        Voice conference number
    #
    # === Return examples (for 0.7)
    #
    # On successful creation:
    #
    #   {
    #    :returncode=>true, :meetingID=>"bigbluebutton-api-ruby-test",
    #    :attendeePW=>"1234", :moderatorPW=>"4321", :hasBeenForciblyEnded=>false,
    #    :messageKey=>"", :message=>""
    #   }
    #
    # Meeting that was forcibly ended:
    #
    #   {
    #    :returncode=>true, :meetingID=>"bigbluebutton-api-ruby-test",
    #    :attendeePW=>"1234", :moderatorPW=>"4321", :hasBeenForciblyEnded=>true,
    #    :messageKey=>"duplicateWarning",
    #    :message=>"This conference was already in existence and may currently be in progress."
    #   }
    #
    def create_meeting(meeting_name, meeting_id, moderator_password = nil, attendee_password = nil,
                       welcome_message = nil, dial_number = nil, logout_url = nil,
                       max_participants = nil, voice_bridge = nil)

      params = { :name => meeting_name, :meetingID => meeting_id,
                 :moderatorPW => moderator_password, :attendeePW => attendee_password,
                 :welcome => welcome_message, :dialNumber => dial_number,
                 :logoutURL => logout_url, :maxParticpants => max_participants,
                 :voiceBridge => voice_bridge }

      response = send_api_request(:create, params)

      response[:meetingID] = response[:meetingID].to_s
      response[:moderatorPW] = response[:moderatorPW].to_s
      response[:attendeePW] = response[:attendeePW].to_s
      response[:hasBeenForciblyEnded] = response[:hasBeenForciblyEnded].downcase == "true"

      response
    end

    # Ends an existing meeting. Throws BigBlueButtonException on failure.
    # meeting_id::          Unique identifier for the meeting
    # moderator_password::  Moderator password
    #
    # === Return examples (for 0.7)
    #
    # On success:
    #
    #   {
    #    :returncode=>true, :messageKey=>"sentEndMeetingRequest",
    #    :message=>"A request to end the meeting was sent.  Please wait a few seconds, and then use the getMeetingInfo
    #               or isMeetingRunning API calls to verify that it was ended."
    #   }
    #
    def end_meeting(meeting_id, moderator_password)
      send_api_request(:end, { :meetingID => meeting_id, :password => moderator_password } )
    end

    # Returns true or false as to whether meeting is open.  A meeting is
    # only open after at least one participant has joined.
    # meeting_id::          Unique identifier for the meeting
    def is_meeting_running?(meeting_id)
      hash = send_api_request(:isMeetingRunning, { :meetingID => meeting_id } )
      hash[:running].downcase == "true"
    end

    # Warning: As of this version of the gem, this call does not work
    # (instead of returning XML response, it should join the meeting).
    #
    # Joins a user into the meeting using an API call, instead of
    # directing the user's browser to moderator_url or attendee_url
    # (note: this will still be required however to actually use bbb).
    # Returns the URL a user can use to enter this meeting.
    # meeting_id::        Unique identifier for the meeting
    # user_name::         Name of the user
    # password::          Moderator or attendee password for this meeting
    # user_id::           Unique identifier for this user
    # web_voice_conf::    Custom voice-extension for users using VoIP
    def join_meeting(meeting_id, user_name, password, user_id = nil, web_voice_conf = nil)
      params = { :meetingID => meeting_id, :password => password, :fullName => user_name,
                 :userID => user_id, :webVoiceConf => web_voice_conf }
      send_api_request(:join, params)
    end

    # Returns a hash object containing the meeting information.
    # See the API documentation for details on the return XML
    # (http://code.google.com/p/bigbluebutton/wiki/API).
    #
    # meeting_id::  Unique identifier for the meeting
    # password::    Moderator password for this meeting
    #
    # === Return examples (for 0.7)
    #
    # With attendees:
    #
    #   {
    #    :returncode=>true, :meetingID=>"bigbluebutton-api-ruby-test", :attendeePW=>"1234", :moderatorPW=>"4321", :running=>true,
    #    :hasBeenForciblyEnded=>false, :startTime=>DateTime("Wed Apr 06 17:09:57 UTC 2011"), :endTime=>nil, :participantCount=>4, :moderatorCount=>2,
    #    :attendees => [
    #      {:userID=>"ndw1fnaev0rj", :fullName=>"House M.D.", :role=>:moderator},
    #      {:userID=>"gn9e22b7ynna", :fullName=>"Dexter Morgan", :role=>:moderator},
    #      {:userID=>"llzihbndryc3", :fullName=>"Cameron Palmer", :role=>:viewer},
    #      {:userID=>"rbepbovolsxt", :fullName=>"Trinity", :role=>:viewer}
    #    ], :messageKey=>"", :message=>""
    #   }
    #
    # Without attendees (not started):
    #
    #   {
    #    :returncode=>true, :meetingID=>"bigbluebutton-api-ruby-test", :attendeePW=>"1234", :moderatorPW=>"4321", :running=>false,
    #    :hasBeenForciblyEnded=>false, :startTime=>nil, :endTime=>nil, :participantCount=>0, :moderatorCount=>0,
    #    :attendees=>[], :messageKey=>"", :message=>""
    #   }
    #
    def get_meeting_info(meeting_id, password)
      response = send_api_request(:getMeetingInfo, { :meetingID => meeting_id, :password => password } )

      # simplify the hash making a node :attendees with an array with all attendees
      if response[:attendees].empty?
        attendees = []
      else
        node = response[:attendees][:attendee]
        if node.kind_of?(Array)
          attendees = node
        else
          attendees = []
          attendees << node
        end
      end
      response[:attendees] = attendees
      response[:attendees].each { |att| att[:role] = att[:role].downcase.to_sym }

      response[:meetingID] = response[:meetingID].to_s
      response[:moderatorPW] = response[:moderatorPW].to_s
      response[:attendeePW] = response[:attendeePW].to_s
      response[:hasBeenForciblyEnded] = response[:hasBeenForciblyEnded].downcase == "true"
      response[:running] = response[:running].downcase == "true"
      response[:startTime] = response[:startTime].downcase == "null" ?
                             nil : DateTime.parse(response[:startTime])
      response[:endTime] = response[:endTime].downcase == "null" ?
                           nil : DateTime.parse(response[:endTime])

      response
    end

    # Returns a hash object containing information about the meetings currently existent in the BBB
    # server, either they are running or not.
    #
    # === Return examples (for 0.7)
    #
    # Server with one or more meetings:
    #
    #   { :returncode => true,
    #     :meetings => [
    #       {:meetingID=>"Demo Meeting", :attendeePW=>"ap", :moderatorPW=>"mp", :hasBeenForciblyEnded=>false, :running=>true},
    #       {:meetingID=>"I was ended Meeting", :attendeePW=>"pass", :moderatorPW=>"pass", :hasBeenForciblyEnded=>true, :running=>false}
    #     ],
    #    :messageKey=>"", :message=>""
    #   }
    #
    # Server with no meetings:
    #
    #   {:returncode=>true, :meetings=>[], :messageKey=>"noMeetings", :message=>"no meetings were found on this server"}
    #
    def get_meetings
      response = send_api_request(:getMeetings, { :random => rand(9999999999) } )

      # simplify the hash making a node :meetings with an array with all meetings
      if response[:meetings].empty?
        meetings = []
      else
        node = response[:meetings][:meeting]
        if node.kind_of?(Array)
          meetings = node
        else
          meetings = []
          meetings << node
        end
      end
      response[:meetings] = meetings

      response[:meetings].each do |meeting|
        meeting[:meetingID] = meeting[:meetingID].to_s
        meeting[:moderatorPW] = meeting[:moderatorPW].to_s
        meeting[:attendeePW] = meeting[:attendeePW].to_s
        meeting[:hasBeenForciblyEnded] = meeting[:hasBeenForciblyEnded].downcase == "true"
        meeting[:running] = meeting[:running].downcase == "true"
      end

      response
    end

    # Returns the API version (as string) of the associated server. This actually returns
    # the version returned by the BBB server, and not the version set by the user in
    # the initialization of this object.
    def get_api_version
      response = send_api_request(:index)
      if response[:returncode]
        response[:version].to_s
      else
        ""
      end
    end

    # Make a simple request to the server to test the connection
    def test_connection
      response = send_api_request(:index)
      response[:returncode]
    end

    # API's are equal if all the following attributes are equal
    def ==(other)
      r = true
      [:url, :supported_versions, :salt, :version, :debug].each do |param|
        r = r && self.send(param) == other.send(param)
      end
      r
    end

    def last_http_response
      @http_response
    end

    protected

    def get_url(method, data)
      if method == :index
        return @url
      end

      url = "#{@url}/#{method}?"

      data.delete_if { |k, v| v.nil? } unless data.nil?
      params = ""
      params = data.map{ |k,v| "#{k}=" + CGI::escape(v.to_s) unless k.nil? || v.nil? }.join("&")

      checksum_param = params + @salt
      checksum_param = method.to_s + checksum_param
      checksum = Digest::SHA1.hexdigest(checksum_param)

      "#{url}#{params}&checksum=#{checksum}"
    end

    def send_api_request(method, data = {})
      url = get_url(method, data)
      begin
        puts "BigBlueButtonAPI: URL request = #{url}" if @debug

        url_parsed = URI.parse(url)
        http = Net::HTTP.new(url_parsed.host, url_parsed.port)
        http.open_timeout = @timeout
        http.read_timeout = @timeout
        @http_response = http.get(url_parsed.path)

        puts "BigBlueButtonAPI: URL response = #{@http_response.body}" if @debug
      rescue TimeoutError => error
        raise BigBlueButtonException.new("Timeout error. Your server is probably down: \"#{@url}\"")
      rescue Exception => error
        raise BigBlueButtonException.new("Connection error. Your URL is probably incorrect: \"#{@url}\"")
      end

      if @http_response.body.empty?
        return { }
      end

      # 'Hashify' the XML
      hash = Hash.from_xml @http_response.body

      # simple validation of the xml body
      unless hash.has_key?(:response) and hash[:response].has_key?(:returncode)
        raise BigBlueButtonException.new("Invalid response body. Is the API URL correct? \"#{@url}\", version #{@version}")
      end

      # and remove the "response" node
      hash = Hash[hash[:response]].inject({}){|h,(k,v)| h[k] = v; h}

      # Adjust some values. There will always be a returncode, message and messageKey in the hash.
      hash[:returncode] = hash[:returncode].downcase == "success"                                  # true instead of "SUCCESS"
      hash[:messageKey] = "" if !hash.has_key?(:messageKey) or hash[:messageKey].empty?            # "" instead of {}
      hash[:message] = "" if !hash.has_key?(:message) or hash[:message].empty?                     # "" instead of {}

      unless hash[:returncode]
        exception = BigBlueButtonException.new(hash[:message])
        exception.key = hash.has_key?(:messageKey) ? hash[:messageKey] : ""
        raise exception
      end

      hash
    end

  end
end


